### PR TITLE
Delay connection closed (#69)

### LINKED ISSUE #74: open_websocket context manager doesn't exit if there are pending receives
is this behaving as intended?

```
import trio
from trio_websocket import ConnectionClosed, serve_websocket, open_websocket_url


class EchoServer:
    """Websocket server that echos any message starting with '?'."""

    def __init__(self):
        self._server = None

    async def _handleConnection(self, request):
        ws = await request.accept()
        while True:
            try:
                message = await ws.get_message()
                if not message.startswith("?"):
                    return
                await ws.send_message(message)
            except ConnectionClosed:
                break

    async def serve(self, port, *, task_status=trio.TASK_STATUS_IGNORED):
        await serve_websocket(self._handleConnection, '127.0.0.1', port,
                              ssl_context=None, task_s```


async def test_foo():
    server = EchoServer()
    async with trio.open_nursery() as nursery:
        await nursery.start(server.serve, 1234)
        async with open_websocket_url('ws://127.0.0.1:1234/') as ws:
            await ws.send_message('?foo')
            # test hangs here because the echo response hasn't been consumed?
        nursery.cancel_scope.cancel()
```

### PR DESCRIPTION
**Pull request by mehaase**
As described in the issue, get_message() was raising connection closed
even if there were pending messages. Per Nathaniel's suggestion, the
proper behavior is this:

1. If the remote endpoint closed the connection, then the local endpoint
   may continue reading all messages sent prior to closing.
2. If the local endpoint closed the connection, then the local endpoint
   may not read any more messages.

I added tests for these two conditions and implemented the behavior by
closing the ReceiveChannel inside the connection's `aclose()`. This
requires a bit of additional exception handling inside `get_message()`
and inside the reader task.

One slight surprise is that the test couldn't be written due to the bug
in #74! The client would hang because the reader task is blocked by
the unconsumed messages. So I changed the channel size to 32, which
allows this test to work, and I will replace this hard-coded value
when I fix #74.

### CODE REVIEW

File: trio_websocket/_impl.py [hunk 1]
```diff
@@ -739,7 +738,13 @@ def subprotocol(self):
         '''
         self._str_message += event.data
         if event.message_finished:
-            await self._send_channel.send(self._str_message)
+            try:
+                await self._send_channel.send(self._str_message)
+            except trio.BrokenResourceError:
```
[the remark this answers is not in the document]
Comment (njsmith) [reply trio_websocket/_impl.py]: Even CPython only makes a "best effort" attempt to keep naive string concatenation cheap – it can totally fail if you get unlucky. Either accumulating in a list or using a `StringIO` should avoid this issue.

[the remark this answers is not in the document]
Comment (belm0) [reply trio_websocket/_impl.py]: I've made the improvements separately in #80

File: tests/test_connection.py [hunk 2]
```diff
@@ -377,6 +377,61 @@ class MemoryListener(trio.abc.Listener):
     with trio.fail_after(2):
         async with open_websocket(
                 HOST, server.port, '/', use_ssl=False) as connection:
-            with pytest.raises(ConnectionClosed) as e:
+            with pytest.raises(ConnectionClosed) as exc_info:
                 await connection.get_message()
-                assert e.reason.name == 'NORMAL_CLOSURE'
+            exc = exc_info.value
```
[the remark this answers is not in the document]
Comment (mehaase) [reply tests/test_connection.py]: I should have mentioned this in the commit message. This assertion was not getting executed because the previous line `await connection.get_message()` was raising an (expected) exception. In order to run this assertion, it needs to happen outside the `pytest.raises` context manager.

File: trio_websocket/_impl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mehaase) [reply trio_websocket/_impl.py]: Thanks, @belm0. I will merge #80 into this branch.

File: trio_websocket/_impl.py [hunk 3]
```diff
@@ -440,12 +442,15 @@ def __init__(self, stream, wsproto, *, path=None):
...
+        self._message_parts = []  # type: List[bytes|str]
         self._reader_running = True
         self._path = path
         self._subprotocol = None
-        self._send_channel, self._recv_channel = trio.open_memory_channel(0)
+        # TODO changed channel size from 0 to 2 temporarily to enable
+        # test_read_messages_after_remote_close to pass. The channel size will
+        # become a configurable setting when #74 is fixed.
+        self._send_channel, self._recv_channel = trio.open_memory_channel(
+            message_queue_size_do_not_use)
```
Comment (belm0) [inline trio_websocket/_impl.py]: I suggest removing this change, and defer adding test_read_messages_after_remote_close until channel size configuration is implemented.  The non-zero size is not needed to fix #74.

So i.e. this PR becomes merely a fix for #74.  In addition to test_no_messages_after_local_close, please add the test from #77.

File: tests/test_connection.py [hunk 4]
```diff
@@ -377,6 +377,61 @@ class MemoryListener(trio.abc.Listener):
...
+            assert exc.reason.name == 'NORMAL_CLOSURE'
+
+
+async def test_read_messages_after_remote_close(nursery):
```
[the remark this answers is not in the document]
Comment (mehaase) [reply tests/test_connection.py]: This is a good point.

File: tests/test_connection.py [hunk 5]
```diff
@@ -377,6 +377,61 @@ class MemoryListener(trio.abc.Listener):
...
+    When the remote endpoint closes, the local endpoint can still read all
+    of the messages sent prior to closing. Any attempt to read beyond that will
+    raise ConnectionClosed.
+    '''
+    server_closed = trio.Event()
+
+    async def handler(request):
+        server = await request.accept()
+        async with server:
+            await server.send_message('1')
```
[the remark this answers is not in the document]
Comment (mehaase) [reply tests/test_connection.py]: I misunderstood what was going on before, but you're right. I will add the `skip` decorator to this test.

File: trio_websocket/_impl.py (hunk 3 above)
Comment (mehaase) [reply trio_websocket/_impl.py]: I'll go for the first suggestion: 

* Set channel size back to zero.
* Add the `skip` decorator to the test that requires non-zero channel size.
* Merge your test from #77.
